applications/CompressiblePotentialFlowApplication/custom_utilities/potential_flow_utilities.cpp

```diff
@@ -506,6 +506,53 @@ double ComputePerturbationLocalMachNumber(const Element& rElement, const Process
     return velocity_module / local_speed_of_sound;
 }
 
+template <int Dim, int NumNodes>
+double ComputeUpwindFactor(
+        const double localMachNumberSquared, 
+        const ProcessInfo& rCurrentProcessInfo)
+{
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), Equation 2.13
+
+    // read free stream values
+    // default MACH_LIMIT - 0.94 and default UPWIND_FACTOR_CONSTANT - 1.0
+    const double critical_mach = rCurrentProcessInfo[MACH_LIMIT];
+    const double upwind_factor_constant = rCurrentProcessInfo[UPWIND_FACTOR_CONSTANT];
+
+    return upwind_factor_constant * (1.0 - std::pow(critical_mach, 2.0) / localMachNumberSquared);
+}
+
+template <int Dim, int NumNodes>
+double SelectMaxUpwindFactor(
+        const array_1d<double, Dim>& rCurrentVelocity, 
+        const array_1d<double, Dim>& rUpwindVelocity, 
+        const ProcessInfo& rCurrentProcessInfo)
+{
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
+    //           and the Integral Boundary Layer Equations in Three Dimensions
+    //           by Brian Nishida (1996), Equation 2.13
+    const double current_element_mach_squared = ComputeLocalMachNumberSquared<Dim,NumNodes>(rCurrentVelocity, rCurrentProcessInfo);
+    const double upwind_element_mach_squared = ComputeLocalMachNumberSquared<Dim,NumNodes>(rUpwindVelocity, rCurrentProcessInfo);
+
+    array_1d<double, 3> upwind_factor_options(3, 0.0);
+    upwind_factor_options[1] = ComputeUpwindFactor<Dim, NumNodes>(current_element_mach_squared, rCurrentProcessInfo);
+    upwind_factor_options[2] = ComputeUpwindFactor<Dim, NumNodes>(upwind_element_mach_squared, rCurrentProcessInfo);
+
+    const auto case_option = ComputeUpwindFactorCase<Dim, NumNodes>(upwind_factor_options);
+    return upwind_factor_options[case_option];
+}
+
+template <int Dim, int NumNodes>
+size_t ComputeUpwindFactorCase(const array_1d<double, 3>& rUpwindFactorOptions)
+{
+    // Following Fully Simulataneous Coupling of the Full Potential Equation
```

Comment:
Could you add a comment explaining each case? Something like:

- Case 0: Subsonic flow
- Case 1: Supersonic and accelerating flow (M^2 > M^2_up)
- Case 2: Supersonic and decelerating flow (M^2 < M^2_up)